test/functional/test_framework/wallet.py

```diff
@@ -370,6 +371,10 @@ def create_self_transfer(
             vsize = Decimal(168)  # P2PK (73 bytes scriptSig + 35 bytes scriptPubKey + 60 bytes other)
         else:
             assert False
+        if target_weight and not fee:  # respect fee_rate if target weight is passed
+            # the actual weight might be off by 3 WUs, so calculate based on that (see self._bulk_tx)
+            max_actual_weight = target_weight + 3
+            fee = get_fee((max_actual_weight + 3) // 4, fee_rate)
         send_value = utxo_to_spend["value"] - (fee or (fee_rate * vsize / 1000))
```

Comment:
```suggestion
        send_value = utxo_to_spend["value"] - (fee or (fee_rate * vsize / 1000))
        # Ensure UTXO is enough to account for the required fee
        assert_greater_than_or_equal(send_value, 0)
```

Reply:
Agreed. However if it had not been for the absolute fee, we could come up with a more generic function `get_effective_unspent_value` that calculates based on the `fee_rate` and the `vsize` of the unspent.

Can't include the absolute fee in this because it is paid for the whole transaction and not just one unspent - still valid for transactions spending only one unspent.